libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp

```diff
@@ -0,0 +1,173 @@
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Common/AP_Common.h>
+#include "AP_BoardConfig_CAN.h"
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_PX4 || CONFIG_HAL_BOARD == HAL_BOARD_VRBRAIN
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+#if HAL_WITH_UAVCAN
+#include <AP_HAL_PX4/CAN.h>
+#endif
+#endif
+
+#if HAL_WITH_UAVCAN
+#include <AP_UAVCAN/AP_UAVCAN.h>
+#endif
+
+extern const AP_HAL::HAL& hal;
+
+// table of user settable parameters
+const AP_Param::GroupInfo AP_BoardConfig_CAN::var_info[] = {
+
+#if HAL_WITH_UAVCAN
+
+#if MAX_NUMBER_OF_CAN_INTERFACES > 0
+    // @Group: P1_
+    // @Path: ../AP_BoardConfig/canbus.cpp
+    AP_SUBGROUPINFO(_var_info_can[0], "P1_", 1, AP_BoardConfig_CAN, AP_BoardConfig_CAN::CAN_if_var_info),
+#endif
+
+#if MAX_NUMBER_OF_CAN_INTERFACES > 1
+    // @Group: P2_
+    // @Path: ../AP_BoardConfig/canbus.cpp
+    AP_SUBGROUPINFO(_var_info_can[1], "P2_", 2, AP_BoardConfig_CAN, AP_BoardConfig_CAN::CAN_if_var_info),
+#endif
+
+#if MAX_NUMBER_OF_CAN_INTERFACES > 2
+    // @Group: P3_
+    // @Path: ../AP_BoardConfig/canbus.cpp
+    AP_SUBGROUPINFO(_var_info_can[2], "P3_", 3, AP_BoardConfig_CAN, AP_BoardConfig_CAN::CAN_if_var_info),
+#endif
+
+#if MAX_NUMBER_OF_CAN_DRIVERS > 0
+    // @Group: D1_
+    // @Path: ../AP_BoardConfig/canbus.cpp
+    AP_SUBGROUPINFO(_var_info_can_protocol[0], "D1_", 4, AP_BoardConfig_CAN, AP_BoardConfig_CAN::CAN_driver_var_info),
+#endif
+
+#if MAX_NUMBER_OF_CAN_DRIVERS > 1
+    // @Group: D2_
+    // @Path: ../AP_BoardConfig/canbus.cpp
+    AP_SUBGROUPINFO(_var_info_can_protocol[1], "D2_", 5, AP_BoardConfig_CAN, AP_BoardConfig_CAN::CAN_driver_var_info),
+#endif
+
+#if MAX_NUMBER_OF_CAN_DRIVERS > 2
+    // @Group: D3_
+    // @Path: ../AP_BoardConfig/canbus.cpp
+    AP_SUBGROUPINFO(_var_info_can_protocol[2], "D3_", 6, AP_BoardConfig_CAN, AP_BoardConfig_CAN::CAN_driver_var_info),
+#endif
+
+#endif
+
+    AP_GROUPEND
+};
+
+#if HAL_WITH_UAVCAN
+int8_t AP_BoardConfig_CAN::_st_driver_number[MAX_NUMBER_OF_CAN_INTERFACES];
+int8_t AP_BoardConfig_CAN::_st_can_debug[MAX_NUMBER_OF_CAN_INTERFACES];
+#endif
+
+void AP_BoardConfig_CAN::init()
+{
+#if HAL_WITH_UAVCAN
+    for (uint8_t i = 0; i < MAX_NUMBER_OF_CAN_INTERFACES; i++)
+    {
+        _st_driver_number[i] = (int8_t) _var_info_can[i]._driver_number;
+        _st_can_debug[i] = (int8_t) _var_info_can[i]._can_debug;
+    }
+#endif
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_PX4 || CONFIG_HAL_BOARD == HAL_BOARD_VRBRAIN
+    px4_setup_canbus();
+#endif
+
+}
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_PX4 || CONFIG_HAL_BOARD == HAL_BOARD_VRBRAIN
+/*
+  setup CANBUS drivers
+ */
+void AP_BoardConfig_CAN::px4_setup_canbus(void)
+{
+#if HAL_WITH_UAVCAN
+    // Create all drivers that we need
+    bool initret = true;
+    for (uint8_t i = 0; i < MAX_NUMBER_OF_CAN_INTERFACES; i++)
+    {
+        uint8_t drv_num = _var_info_can[i]._driver_number;
+
+        if (drv_num <= MAX_NUMBER_OF_CAN_DRIVERS) {
```

Comment:
Now you are not checking it isn't zero :slightly_smiling_face: We should actually test it is bigger than zero.